habitat_baselines/common/utils.py

```diff
@@ -0,0 +1,148 @@
+import argparse
+import glob
+import os
+from collections import defaultdict
+from typing import Optional
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from habitat import Config, get_config, make_dataset
+from habitat_baselines.config.default import get_config as baseline_cfg
+
+
+class Flatten(nn.Module):
+    def forward(self, x):
+        return x.view(x.size(0), -1)
+
+
+def _flatten_helper(t, n, tensor):
+    return tensor.view(t * n, *tensor.size()[2:])
+
+
+class CustomFixedCategorical(torch.distributions.Categorical):
+    def sample(self, sample_shape=torch.Size()):
+        return super().sample(sample_shape).unsqueeze(-1)
+
+    def log_probs(self, actions):
+        return (
+            super()
+            .log_prob(actions.squeeze(-1))
+            .view(actions.size(0), -1)
+            .sum(-1)
+            .unsqueeze(-1)
+        )
+
+    def mode(self):
+        return self.probs.argmax(dim=-1, keepdim=True)
+
+
+class CategoricalNet(nn.Module):
+    def __init__(self, num_inputs, num_outputs):
+        super().__init__()
+
+        self.linear = nn.Linear(num_inputs, num_outputs)
+
+        nn.init.orthogonal_(self.linear.weight, gain=0.01)
+        nn.init.constant_(self.linear.bias, 0)
+
+    def forward(self, x):
+        x = self.linear(x)
+        return CustomFixedCategorical(logits=x)
+
+
+def _flatten_helper(t, n, tensor):
+    return tensor.view(t * n, *tensor.size()[2:])
+
+
+def update_linear_schedule(optimizer, epoch, total_num_epochs, initial_lr):
+    r"""Decreases the learning rate linearly
+    """
+    lr = initial_lr - (initial_lr * (epoch / float(total_num_epochs)))
+    for param_group in optimizer.param_groups:
+        param_group["lr"] = lr
+
+
+def experiment_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--run-type",
+        choices=["train", "eval"],
+        required=True,
+        help="run type of the experiment (train or eval)",
+    )
+    parser.add_argument(
+        "--exp-config",
+        type=str,
+        required=True,
+        help="path to config yaml containing info about experiment",
+    )
+    parser.add_argument(
+        "opts",
+        default=None,
+        nargs=argparse.REMAINDER,
+        help="Modify config options from command line",
+    )
+    return parser
+
+
+def get_exp_config(cfg_path, opts=None):
+    config = Config(new_allowed=True)
```

Comment:
How `new_allowed`  works?

Reply:
`new_allowed=True` allows merging configs even if they have non-overlapping config nodes. For example, merging baseline_config into env_config with `new_allowed=False` won't work since env_config by default doesn't have the `BASELINE` node.

Reply:
@JasonJiazhiZhang is that new yacs feature? It wasn't there before.

Reply:
@mathfac I don't think it's new. See [here](https://github.com/rbgirshick/yacs/blob/aabe5bd54d193996ceedcff650cff10c38b1f90b/yacs/config.py#L74).